File: quote/Projects/PluginModel/ModelEvents.vb
Imports System.ComponentModel

Imports Model.Common

Public Class ModelEventArgs 
    Inherits EventArgs 
    
    Public Property Id
End Class

Public NotInheritable Class ModelEvents

    Private Shared _ActiveEvents As ModelEvents = New ModelEvents

    Public Shared Event TemplateCreated As EventHandler(Of ModelEventArgs)

    Public Shared Event TemplateViewed As EventHandler

    Public Shared Event QuoteViewed As EventHandler

    Private Sub New
        
    End Sub
    
    Public Shared ReadOnly Property ActiveEvents As ModelEvents 
        Get
            Return _ActiveEvents
        End Get
    End Property
    
    Public Shared Sub NotifyTemplateCreated(id As Integer)
        Dim args As New ModelEventArgs
        args.Id = id
        RaiseEvent TemplateCreated(Nothing, args)
    End Sub

    Public Shared Sub NotifyTemplateViewed()
        RaiseEvent TemplateViewed(Nothing, New EventArgs())
    End Sub

    Public Shared Sub NotifyQuoteViewed()
        RaiseEvent QuoteViewed(Nothing, New EventArgs())
    End Sub

End Class

File: quote/Projects/PluginModel/ModelEvents.cs
namespace Model
{
    using System;
    using System.Linq;
    
    /// <summary>
    /// Used to signal changes to the model.  Like when a quote is loaded.
    /// This is a singleton class.
    /// </summary>
    public sealed class ModelEvents
    {
        /// <summary>
        /// The only instance.
        /// </summary>
        private static ModelEvents instance = new ModelEvents();
        
        /// <summary>
        /// Prevents a default instance of the <see cref="ModelEvents"/> class from being created.
        /// </summary>
        private ModelEvents()
        {
            // Singleton class
        }

        /// <summary>
        /// Signals when a template is created.
        /// </summary>
        public event EventHandler<ModelEventArgs> TemplateCreated;

        /// <summary>
        /// Signals when a template is viewed.
        /// </summary>
        public event EventHandler TemplateViewed;

        /// <summary>
        /// Signals when a quote is viewed.
        /// </summary>
        public event EventHandler QuoteViewed;

        /// <summary>
        /// Gets the instance.
        /// </summary>
        /// <value>The static instance.</value>
        public static ModelEvents Instance
        {
            get { return instance; }
        }

        /// <summary>
        /// Signal that a template has been created.
        /// </summary>
        /// <param name="id">The id of the model object.</param>
        public void NotifyTemplateCreated(int id)
        {
            ModelEventArgs args = new ModelEventArgs();
            args.Id = id;
            if (this.TemplateCreated != null) 
            {
                this.TemplateCreated(null, args);
            }
        }

        /// <summary>
        /// Signal that a template has been viewed.
        /// </summary>
        public void NotifyTemplateViewed()
        {
            if (this.TemplateViewed != null) 
            {
                this.TemplateViewed(null, new EventArgs());
            }
        }

        /// <summary>
        /// Signal that a quote has been viewed.
        /// </summary>
        public void NotifyQuoteViewed()
        {
            if (this.QuoteViewed != null) 
            {
                this.QuoteViewed(null, new EventArgs());
            }
        }
    }
}